lib/AST/ASTContext.cpp

```diff
@@ -2212,6 +2214,8 @@ ASTContext::getModule(ImportPath::Module ModulePath) {
   auto moduleID = ModulePath[0];
   for (auto &importer : getImpl().ModuleLoaders) {
     if (ModuleDecl *M = importer->loadModule(moduleID.Loc, ModulePath)) {
+      if (callback)
+        callback(llvm::formatv("Loading module {0}", M->getNameStr()).str());
```

Comment:
It could be nice if the module loaders could customize the message to indicate that e.g., a module is being compiled from source.
Do we need to change the callback for subsequent invocations? If not, should maybe each module loader's constructor take an optional callback parameter instead of loadModule()?

Reply:
> It could be nice if the module loaders could customize the message to indicate that e.g., a module is being compiled from source.

I agree, but I'm thinking the module loaders could just have static string with a getter that would be used by the callback function.

> Do we need to change the callback for subsequent invocations? If not, should maybe each module loader's constructor take an optional callback parameter instead of loadModule()?

I guess you meant `getModule` not `loadModule`, right? If it's so, when `getModule` with call the module loader `loadModule` implementation, since it's virtual pure, we would need to change each module loader implementation of `loadModule` to make sure it calls the callback. To me it's safer to invoke the callback at the `ASTContext` level.